### PR TITLE
Changes to allow libunwind to be linked to libstd of rust-sgx.

### PR DESCRIPTION
**Pull request by VardhanThigle**
# Initial Changes to allow libunwind to be linked to libstd of rust-sgx.
### Intial Fork
* Inital Fork has been made from 5.0 release of libunwind (commit: 29d162446817)
* git diftool 29d162446817     should show the changes done so far.
#### Libraries we do not link to current target
 * pthread

### Clone the LLVM source tree
* `git clone https://github.com/llvm-mirror/llvm.git`
### Generate Unix Makefiles
* `cd build`
* `cmake -DCMAKE_BUILD_TYPE="RELEASE" -DRUST_SGX=1 -G "Unix Makefiles" -DLLVM_PATH=../../llvm/ ../`
* you could remove `-DCMAKE_BUILD_TYPE="RELEASE"` to enable debug logs of libunwind.
### Build!!
* `make unwind_static`
     build will be found at build/lib/
### Clen build
* cd `build` directory.
* rm -rf ./*
* touch ./.keep

### CODE REVIEW

**Comment by jethrogb** [comment]
Can you also put a link to the corresponding changes to the rust repo?

File: CMakeLists.txt [hunk 1]
```diff
@@ -272,7 +272,7 @@ if (LIBUNWIND_ENABLE_ASSERTIONS)
 
   # On Release builds cmake automatically defines NDEBUG, so we
   # explicitly undefine it:
-  if (uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
+  if ((uppercase_CMAKE_BUILD_TYPE STREQUAL "RELEASE") AND (NOT RUST_SGX))
```
Comment (jethrogb) [inline CMakeLists.txt:275]: Is this necessary?

File: src/AddressSpace.hpp [hunk 2]
```diff
@@ -14,7 +14,9 @@
 #define __ADDRESSSPACE_HPP__
 
 #include <stdint.h>
+#ifndef RUST_SGX
 #include <stdio.h>
```
Comment (jethrogb) [inline src/AddressSpace.hpp]: Which definitions are we avoiding this way?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -44,6 +45,26 @@ if (MSVC_IDE)
...
+	# Compile Flags
+	add_definitions(-DRUST_SGX)
+	add_definitions(-DU_FORTIFY_SOURCE)
+	add_definitions(-D_FORTIFY_SOURCE=0)
+	add_definitions(-D__NO_STRING_INLINES)
+	add_definitions(-D__NO_MATH_INLINES)
+	add_definitions(-D_LIBUNWIND_IS_BAREMETAL)
+	list(APPEND LIBUNWIND_COMPILE_FLAGS -fno-stack-protector)
+        list(APPEND LIBUNWIND_COMPILE_FLAGS -ffreestanding)
+        list(APPEND LIBUNWIND_COMPILE_FLAGS -fexceptions)
```
Comment (jethrogb) [inline src/CMakeLists.txt]: Indentation issue

File: src/libunwind.cpp [hunk 4]
```diff
@@ -350,38 +350,23 @@ void _unw_remove_dynamic_fde(unw_word_t fde) {
...
-  static bool checked = false;
-  static bool log = false;
-  if (!checked) {
-    log = (getenv("LIBUNWIND_PRINT_APIS") != NULL);
-    checked = true;
-  }
-  return log;
+  // No get env support for rust binding yet
+  // In sgx we might not want to start logging via an env setting.
+  return true;
```
Comment (jethrogb) [inline src/libunwind.cpp:356]: Use `#ifdef RUST_SGX`

File: src/UnwindRustSgx.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+int
+dl_iterate_phdr (int (*callback) (struct dl_phdr_info *,
+				  size_t, void *),
+		 void *data)
+{
+	struct dl_phdr_info info;
+	struct dl_phdr_info *pinfo = &info;
+	int ret = 0;
+
+        Elf_Phdr phdr[2];
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: Indentation issue

File: src/UnwindRustSgx.c [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+#include "UnwindRustSgx.h"
+
+#ifdef __cplusplus
+extern "C"  {
+#endif
+	int c_rwlock_rdlock(RWLock *rwlock);
+	int c_rwlock_wrlock(RWLock *rwlock);
+	int c_rwlock_unlock(RWLock *rwlock);
+	unsigned char *__rust_alloc(size_t, size_t);
+	void __rust_dealloc(unsigned char *, size_t);
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: I don't think this is the correct signature for this function

File: src/UnwindRustSgx.c [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	phdr[1].p_memsz = eh_base_size;
+
+
+	ret = callback (&info, sizeof (struct dl_phdr_info), data);
+	return ret;
+}
+
+void *libuw_malloc(size_t size)
+{
+	return (void *)__rust_alloc(size, 1);
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: `__rust_dealloc` requires that you pass the size. `malloc` normally uses a minimum alignent of `sizeof(size_t)`

File: src/UnwindRustSgx.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+#define stderr NULL
+
+
+
+typedef struct {
+	unsigned char opaque[128];
+} RWLock;
+
+#define RWLOCK_INIT { \
```
Comment (jethrogb) [inline src/UnwindRustSgx.h:31]: Add a comment explaining where this comes from

File: src/UnwindRustSgx.c [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	void __rust_dealloc(unsigned char *, size_t);
+	int c_print_msg(uint8_t *m, int s);
+	void c_abort(void);
+	unsigned char *c_encl_address(size_t);
+
+#ifdef __cplusplus
+}
+#endif
+
+void abort() __attribute__((noreturn)) {
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: For the following functions:
* abort
* libuw_pthread_rwlock_rdlock
* libuw_pthread_rwlock_wrlock
* libuw_pthread_rwlock_unlock
* __libuw_stack_chk_fail

Can you just use `#define` instead of writing all these forwarding implementations? Or just name the functions correctly in Rust?

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -44,6 +45,26 @@ if (MSVC_IDE)
...
+
+	# Sources
+	list(APPEND LIBUNWIND_C_SOURCES UnwindRustSgx.c)
+	list(APPEND LIBUNWIND_C_SOURCES UnwindRustSgxSnprintf.c)
```
Comment (jethrogb) [inline src/CMakeLists.txt]: Is this only necessary if debugging is enabled?

File: src/UnwindRustSgx.h [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+#ifdef RUST_SGX
+
+#define _GNU_SOURCE
+#include <link.h>
+#include <stdarg.h>
+#include <string.h>
+#include <stdint.h>
+
+
+#define stderr NULL
```
Comment (jethrogb) [inline src/UnwindRustSgx.h]: Maybe use a different sentinel value to avoid mistakes when calling `fprintf` etc.

File: src/UnwindRustSgx.c [hunk 12]
```diff
@@ -0,0 +1,150 @@
+#define _GNU_SOURCE
+#include <link.h>
+
+#include <elf.h>
+#include <stdarg.h>
+#define max_log 256
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: Don't put this define between some includes

**Review by jethrogb — CHANGES_REQUESTED** [review]
More comments to come once I take a look at the Rust code

File: CMakeLists.txt (hunk 1 above)
Comment (VardhanThigle) [reply CMakeLists.txt:275]: This is to suppress debug prints in release build. 

Actually, Cmake env by default, sets NDEBUG for release build, and the libunwind has compile time checks for NDEBUG for printing or not debug logs. But, they also have a framework of enabling logs through env variables, which they want to retain in release build, hence by default CmakeList undefines the NDEBUG for release builds.

Since, we want to have a compile time switch for debug logs, I thought that not disabling NDEBUG for release build might be the appropriate way. Please let me know if we want to go about this any other way.

File: src/AddressSpace.hpp (hunk 2 above)
Comment (VardhanThigle) [reply src/AddressSpace.hpp]: This was added mostly due to stderr. Also we redefine fprintf to our own function.
But I can try without removing stdout too, it might make it a rule to include UnwindRustSgx.h to be included after stdio.h.

File: src/CMakeLists.txt (hunk 3 above)
Comment (VardhanThigle) [reply src/CMakeLists.txt]: Will fix.

File: src/libunwind.cpp (hunk 4 above)
Comment (VardhanThigle) [reply src/libunwind.cpp:356]: Thanks. will fix this.

File: src/UnwindRustSgx.h (hunk 8 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.h:31]: Sure.

File: src/UnwindRustSgx.c (hunk 5 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.c]: Will fix.

File: src/UnwindRustSgx.c (hunk 12 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.c]: will fix.

File: src/UnwindRustSgx.h (hunk 8 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.h:31]: Sure.

File: src/UnwindRustSgx.c (hunk 6 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.c]: Thanks, Jethro. Will fix.

Comment (VardhanThigle) [reply src/UnwindRustSgx.c]: Thanks. Will fix

File: src/CMakeLists.txt (hunk 3 above)
Comment (VardhanThigle) [reply src/CMakeLists.txt]: Sorry, just realized after compiling, that there is a print that comes in Release build in case of an assert, which needs snprintf.

File: src/CMakeLists.txt [hunk 13]
```diff
@@ -9,16 +9,10 @@ set(LIBUNWIND_C_SOURCES
...
-set_source_files_properties(${LIBUNWIND_C_SOURCES}
-                            PROPERTIES
-                              COMPILE_FLAGS "-std=c99")
 
 set(LIBUNWIND_ASM_SOURCES
     UnwindRegistersRestore.S
     UnwindRegistersSave.S)
-set_source_files_properties(${LIBUNWIND_ASM_SOURCES}
-                            PROPERTIES
-                              LANGUAGE C)
```
Comment (jethrogb) [inline src/CMakeLists.txt:21]: Why is this moved?

Comment (VardhanThigle) [reply src/CMakeLists.txt:21]: I realized that ASM sources we append which is MEM.S were not getting their properties set. So thought it is more correct to call set_source_file_properties after the list is completely formed before the list is merged into another.
Mem.S was not present in the archive and we had memset & memcpy as Undefined function.

Please Let me know if any other way  would be more appropriate

File: src/UnwindRustSgx.c [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    Elf_Phdr phdr[2];
+    int ret = 0;
+
+
+    size_t text_size = ENCLAVE_TEXT_SIZE;
+    size_t eh_base_size = EH_FRM_HDR_SIZE;
+
+    memset(pinfo, 0, sizeof(*pinfo));
+
+    pinfo->dlpi_addr = (void *)__rust_encl_address(0);
```
Comment (jethrogb) [inline src/UnwindRustSgx.c]: Just use `extern void * IMAGE_BASE`, then you can remove the `__rust_encl_address` function

File: src/CMakeLists.txt (hunk 10 above)
Comment (VardhanThigle) [reply src/CMakeLists.txt]: This is needed for printing assert conditions in release build.

File: src/UnwindRustSgx.c (hunk 14 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.c]: Thanks. Will use IMAGE_BASE

File: src/UnwindRustSgx.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+#include <stdint.h>
+
+// We have to use RWLock from rust repo, it is defined in:
+// src/libstd/sys/sgx/rwlock.rs.
+// rwlock.rs has compile time check to ensure sizeof(RWLock) = 128.
+typedef struct {
+    unsigned char opaque[128];
+} RWLock;
+
+// The below is obtained by printing initilizated bytes
```
Comment (jethrogb) [inline src/UnwindRustSgx.h]: Typo: initialized

File: src/UnwindRustSgx.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+    0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+    0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+    0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+    0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, \
+}
+
+// These are the functions exposed by SGX-Rust.
+// The rust changes are available at:
+// https://github.com/VardhanThigle/rust/tree/Vardhan/rust-sgx-unwind-support
```
Comment (jethrogb) [inline src/UnwindRustSgx.h]: Remove URL in comment

File: src/UnwindRustSgx.h [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+#define PTHREAD_RWLOCK_INITIALIZER RWLOCK_INIT
+
+
+#define pthread_rwlock_rdlock __rust_rwlock_rdlock
+#define pthread_rwlock_wrlock __rust_rwlock_wrlock
+#define pthread_rwlock_unlock __rust_rwlock_unlock
+
+
+#define malloc libuw_malloc
+#define free libuw_free
```
Comment (jethrogb) [inline src/UnwindRustSgx.h:84]: If you use weak linking instead of renaming, does that allow you to get rid of most of the `#include "UnwindRustSgx.h"` statements?

File: src/UnwindRustSgxMem.S [hunk 18]
```diff
@@ -0,0 +1,57 @@
+// (c) Fortanix, Inc.
```
Comment (jethrogb) [inline src/UnwindRustSgxMem.S]: I'm fairly certain I already commented on this, but I can't find it. I don't think this file should be necessary. memcpy/memmove/memset should be provided by compiler_builtins

**Review by jethrogb — CHANGES_REQUESTED** [review]
Approved with minor changes

File: src/UnwindRustSgx.h (hunk 17 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.h:84]: Done. Actually we needed it in config.h. That made it quite simpler.

File: src/UnwindRustSgx.h (hunk 16 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.h]: Fixed. Pushing the changes

File: src/UnwindRustSgx.h (hunk 15 above)
Comment (VardhanThigle) [reply src/UnwindRustSgx.h]: Fixed. Pushing the changes

File: src/UnwindRustSgxMem.S (hunk 18 above)
Comment (VardhanThigle) [reply src/UnwindRustSgxMem.S]: Fixed. Pushing the changes

**Comment by jethrogb** [comment]
Add license header in new files, e.g.:
```
//===--------------------- Unwind_AppleExtras.cpp -------------------------===//
//
//                     The LLVM Compiler Infrastructure
//
// This file is dual licensed under the MIT and the University of Illinois Open
// Source Licenses. See LICENSE.TXT for details.
//
//
//===----------------------------------------------------------------------===//
```
Make sure to fix the filename

**Review by jethrogb — APPROVED** [review]

**Comment by jethrogb** [comment]
Please squash changes into single commit before merging.
